sklearn/utils/estimator_checks.py

```diff
@@ -821,24 +822,25 @@ def _is_pairwise_metric(estimator):
     return bool(metric == "precomputed")
 
 
-def _generate_sparse_matrix(X_csr):
-    """Generate sparse matrices with {32,64}bit indices of diverse format.
+def _generate_sparse_data(X_csr):
+    """Generate sparse matrices or arrays with {32,64}bit indices of diverse format.
 
     Parameters
     ----------
-    X_csr: CSR Matrix
-        Input matrix in CSR format.
+    X_csr: scipy.sparse.csr_matrix or scipy.sparse.csr_array
+        Input in CSR format.
 
     Returns
     -------
-    out: iter(Matrices)
+    out: iter(Matrices) or iter(Arrays)
         In format['dok', 'lil', 'dia', 'bsr', 'csr', 'csc', 'coo',
         'coo_64', 'csc_64', 'csr_64']
     """
 
     assert X_csr.format == "csr"
     yield "csr", X_csr.copy()
-    for sparse_format in ["dok", "lil", "dia", "bsr", "csc", "coo"]:
+    # TODO re-insert "dia" when PR #27372 is merged
+    for sparse_format in ["dok", "lil", "bsr", "csc", "coo"]:
         yield sparse_format, X_csr.asformat(sparse_format)
```

Comment:
Here, for the "coo" case, we need to make sure that the indices are always 32-bits. Matrices will automatically downcase but this is not the case of array. This is probably the failure that we see in the `Lasso` and `ElasticNet`:

Reply:
Hm, if I understand correctly, then large sparse matrices would only downcast automatically, if the data has little enough non-zero values so that their indices fit into the 32 bits, otherwise it would use 64 bits?

The estimators in scikit-learn, that currently `accept_large_sparse=True`  won't fail if we keep it as is here.

Except for Lasso and ElasticNet, that in fact don't support large data, not even for sparse matrices. I would change that in `check_array` of the estimators, not here.

Reply:
I'm also not sure why we would cast here @glemaitre . If anything, I feel like the fact that `.asformat` downcasts automatically for spmatrix, is making us miss the cases where estimators don't support large sparse, but we think we're testing them.

Reply:
I things that there 2 things here:

- in the first loop `for sparse_format in [...]`, we intend to test the 32-bits behaviour and we need to cast the indices because #27372 intend to impose the same behaviour between matrix and array (and this is already the case in scipy-dev since we reported the regression).
- the case under the `for` loop is especially crafted for 64-bits indices. This case is the one that is testing if we support the large indices.

On the top of that I think that we need to expand this 64-bits to CSR/CSC format in addition to COO because it will be more common that a COO matrix will be converted to CSR or CSC. Thus, we can these different format to see what happen when the matrix is not going to be converted.

Reply:
Thank you, @glemaitre. 

I have re-inserted the "dia" sparse data format into the data generator. Am I understanding correctly that with the merge of #27372 nothings else needs to be changed here?

There are still two issues from my side:

The CI had failed because of a DeprecationWarning from SciPy, which I have handled with `ignore_warnings` (though not sure when to use `warnings.filterwarnings` instead).

Also, regarding the code coverage CI failure, I am not sure why it is complaining, because it seems to me that this PR doesn't cause fewer lines to be tested. Should I do something here?

Reply:
I will make a review and check those.